internal/configs/checks.go

```diff
@@ -32,6 +34,27 @@ type CheckRule struct {
 	DeclRange hcl.Range
 }
 
+// validateSelfReferences looks for references in the check rule matching the
+// specified resource address, returning error diagnostics if such a reference
+// is found.
+func (cr *CheckRule) validateSelfReferences(checkType string, addr addrs.Resource) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+	refs, _ := lang.References(cr.Condition.Variables())
+	addrStr := addr.String()
+	for _, ref := range refs {
+		if ref.Subject.String() == addrStr {
```

Comment:
I think `lang.References` prefers to return `addrs.ResourceInstance` where possible, such as if the user literally wrote out `aws_instance.foo[0].bar` then I would expect this string to be `aws_instance.foo[0]` rather than just `aws_instance.foo` and thus not match.

For target addresses (`addrs.Targetable`) we have the idea of asking if one is contained inside another, but I guess we don't have any comparable thing for `addrs.Referenceable`, but probably this doesn't need to be so generic anyway and it could instead be a specialized check to peel off the instance wrapper if it's present, since we know we only care about resources here anyway?

```go
refAddr := ref.Subject
if instAddr, ok := refAddr.(addrs.ResourceInstance); ok {
    refAddr = refAddr.Resouce
}
if refAddr.String() == addrStr {
    // ...
```

Reply:
Oops, good spot, thanks! I've done essentially what you suggested here but with a type switch. Also noticed that I made the same compare-addrs-as-strings mistake again so I've updated the comparison to use `.Equal()` too.